File: PHP/One/Transforms/InferTypesPlugins/InferForeachVarType.php
<?php

namespace One\Transforms\InferTypesPlugins\InferForeachVarType;

use One\Transforms\InferTypesPlugins\Helpers\InferTypesPlugin\InferTypesPlugin;
use One\Ast\AstTypes\ClassType;
use One\Ast\AstTypes\InterfaceType;
use One\Ast\AstTypes\IInterfaceType;
use One\Ast\AstTypes\AnyType;
use One\Ast\Statements\Statement;
use One\Ast\Statements\ForeachStatement;

class InferForeachVarType extends InferTypesPlugin {
    function __construct() {
        parent::__construct("InferForeachVarType");
        
    }
    
    function handleStatement($stmt) {
        if ($stmt instanceof ForeachStatement) {
            $stmt->items = $this->main->runPluginsOn($stmt->items);
            $arrayType = $stmt->items->getType();
            $found = false;
            if ($arrayType instanceof ClassType || $arrayType instanceof InterfaceType) {
                $intfType = $arrayType;
                $isArrayType = \OneLang\ArrayHelper::some($this->main->currentFile->arrayTypes, function ($x) use ($intfType) { return $x->decl === $intfType->getDecl(); });
                if ($isArrayType && count($intfType->typeArguments) > 0) {
                    $stmt->itemVar->type = $intfType->typeArguments[0];
                    $found = true;
                }
            }
            
            if (!$found && !($arrayType instanceof AnyType))
                $this->errorMan->throw("Expected array as Foreach items variable, but got " . $arrayType->repr());
            
            $this->main->processBlock($stmt->body);
            return true;
        }
        return false;
    }
}

File: CSharp/One/Transforms/InferTypesPlugins/InferForeachVarType.cs
using One.Ast;
using One.Transforms.InferTypesPlugins.Helpers;

namespace One.Transforms.InferTypesPlugins
{
    public class InferForeachVarType : InferTypesPlugin {
        public InferForeachVarType(): base("InferForeachVarType")
        {
            
        }
        
        public override bool handleStatement(Statement stmt)
        {
            if (stmt is ForeachStatement forStat) {
                forStat.items = this.main.runPluginsOn(forStat.items);
                var arrayType = forStat.items.getType();
                var found = false;
                if (arrayType is ClassType classType || arrayType is InterfaceType) {
                    var intfType = ((IInterfaceType)arrayType);
                    var isArrayType = this.main.currentFile.arrayTypes.some(x => x.decl == intfType.getDecl());
                    if (isArrayType && intfType.typeArguments.length() > 0) {
                        forStat.itemVar.type = intfType.typeArguments.get(0);
                        found = true;
                    }
                }
                
                if (!found && !(arrayType is AnyType))
                    this.errorMan.throw_($"Expected array as Foreach items variable, but got {arrayType.repr()}");
                
                this.main.processBlock(forStat.body);
                return true;
            }
            return false;
        }
    }
}